File: Semana 5/actividad 3.4/main.py


class Tiempo:
    def __init__(self, hora=0, minuto=0, segundo=0):
        self._hora = hora
        self._minuto = minuto
        self._segundo = segundo

    def set_hora(self, h):
        if 0 <= h < 24:
            self._hora = h
        else:
            self._hora = 0
        return self   # Retorna self para encadenar

    def set_minuto(self, m):
        if 0 <= m < 60:
            self._minuto = m
        else:
            self._minuto = 0
        return self

    def set_segundo(self, s):
        if 0 <= s < 60:
            self._segundo = s
        else:
            self._segundo = 0
        return self

    def mostrar(self):
        print(f"{self._hora}:{self._minuto}:{self._segundo}")

# Uso
t = Tiempo()
t.set_hora(10).set_minuto(30).set_segundo(15)
t.mostrar()   # 10:30:15

# También se puede asignar el resultado (aunque es el mismo objeto)
t2 = t.set_hora(8).set_minuto(45).set_segundo(0)
t2.mostrar()   # 8:45:0
t.mostrar()    # 8:45:0 (t y t2 son el mismo objeto)

File: Semana 5/actividad 3.4/main.cpp

#include <iostream>

class Tiempo {
private:
    int hora, minuto, segundo;

public:
    Tiempo(int h = 0, int m = 0, int s = 0) : hora(h), minuto(m), segundo(s) {}

    // Cada método set retorna una referencia al objeto actual (*this)
    Tiempo& setHora(int h) {
        if (h >= 0 && h < 24)
            hora = h;
        else
            hora = 0;
        return *this;   // Devuelve el objeto actual por referencia
    }

    Tiempo& setMinuto(int m) {
        if (m >= 0 && m < 60)
            minuto = m;
        else
            minuto = 0;
        return *this;
    }

    Tiempo& setSegundo(int s) {
        if (s >= 0 && s < 60)
            segundo = s;
        else
            segundo = 0;
        return *this;
    }

    void mostrar() const {
        std::cout << hora << ":" << minuto << ":" << segundo << std::endl;
    }
};

int main() {
    Tiempo t;

    // Encadenamiento de llamadas
    t.setHora(10).setMinuto(30).setSegundo(15);
    t.mostrar();   // 10:30:15

    // También se puede encadenar la propia muestra si mostrar retornara *this,
    // pero no es necesario.

    return 0;
}
